src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -0,0 +1,294 @@
+                $userPrefixString
+            );
+        }
+
+        return $message;
+    }
+
+
+    private function generateSlugData(array $lineItems): void
+    {
+        $lineItemSlugs = [];
+        foreach ($lineItems as $lineItem) {
+            $lineItemSlugs[$lineItem->getId()] = $lineItem->getSlug();
+        }
+        $this->lineItemSlugs = $lineItemSlugs;
+    }
+
+    private function createBulkUserAssignmentData(
+        array $userNameLastIndexes,
+        array $userGroupIds,
+        array $userPrefix,
+        int $userGroupAssignCount,
+        ?string $groupPrefix,
+        int $batchSize
+    ): array {
+        $slugWiseTotalUsersArray = [];
+        $automateCsvPath = $this->generatedUsersFilePath . date('Y-m-d');
+
+        if (!$this->filesystem->exists($automateCsvPath)) {
+            $this->filesystem->mkdir($automateCsvPath);
+        }
+
+        foreach ($userPrefix as $prefix) {
+            $csvPath = sprintf('%s/%s', $automateCsvPath, $prefix);
+            if (!$this->filesystem->exists($csvPath)) {
+                $this->filesystem->mkdir($csvPath);
+            }
+
+            foreach ($this->lineItemSlugs as $lineSlugs) {
+                $slugTotalUsersCreated = 0;
+                $csvFilename = sprintf('%s-%s.csv', $lineSlugs, $prefix);
+                $csvData = $assignmentCsvData = [];
+                foreach (range(1, $batchSize) as $batchIncrementValue) {
+                    $username = sprintf(
+                        '%s_%s_%d',
+                        $lineSlugs,
+                        $prefix,
+                        (
+                            (int)$userNameLastIndexes[$lineSlugs] + $batchIncrementValue
+                        )
+                    );
+                    $userPassword = $this->createUserPassword();
+                    $userGroupId = $groupPrefix
+                        ? $this->createUserGroupId($userGroupIds, $userGroupAssignCount)
+                        : '';
+
+                    $csvData[] = [$username, $userPassword,$userGroupId];
+                    $assignmentCsvData[] = [$username, $lineSlugs];
+
+                    $slugTotalUsersCreated++;
+                }
+
+                $this->csvWriter->writeCsvData(
+                    $lineSlugs,
+                    $prefix,
+                    $csvPath,
+                    $csvFilename,
+                    $automateCsvPath,
+                    $csvData,
+                    $assignmentCsvData
+                );
+
+                $slugWiseTotalUsersArray[$lineSlugs] = array_key_exists($lineSlugs, $slugWiseTotalUsersArray)
+                    ? $slugWiseTotalUsersArray[$lineSlugs] +  $slugTotalUsersCreated
+                    : $slugTotalUsersCreated;
+            }
+        }
+
+        return $slugWiseTotalUsersArray;
+    }
+
+    private function createUserPassword(): string
+    {
+        return substr(str_shuffle('0123456789abcdefghijklmnopqrstvwxyz'), 0, 8);
```

Comment:
that could be a weak password. is that ok?

Reply:
Yep, in general that's ok for users we are aiming at ;) 
p.s.
One of the latter stages will most probably include possibility to either configure password generation, or use the strategy already configured in security.yaml